crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -1105,33 +1355,41 @@ impl<P: PhaseItem, const I: usize> RenderCommand<P> for SetMeshViewBindGroup<I>
 
 pub struct SetMeshBindGroup<const I: usize>;
 impl<P: PhaseItem, const I: usize> RenderCommand<P> for SetMeshBindGroup<I> {
-    type Param = SRes<MeshBindGroup>;
+    type Param = SRes<PreparedMeshes>;
     type ViewWorldQuery = ();
     type ItemWorldQuery = (
         Read<DynamicUniformIndex<MeshUniform>>,
         Option<Read<SkinnedMeshJoints>>,
+        Option<Read<GpuMorphTargets>>,
     );
     #[inline]
     fn render<'w>(
         _item: &P,
         _view: (),
-        (mesh_index, skinned_mesh_joints): ROQueryItem<'_, Self::ItemWorldQuery>,
-        mesh_bind_group: SystemParamItem<'w, '_, Self::Param>,
+        (mesh_index, skinned_mesh_joints, morph_targets): ROQueryItem<'_, Self::ItemWorldQuery>,
+        prepared_meshes: SystemParamItem<'w, '_, Self::Param>,
         pass: &mut TrackedRenderPass<'w>,
     ) -> RenderCommandResult {
-        if let Some(joints) = skinned_mesh_joints {
-            pass.set_bind_group(
-                I,
-                mesh_bind_group.into_inner().skinned.as_ref().unwrap(),
-                &[mesh_index.index(), joints.index],
-            );
-        } else {
-            pass.set_bind_group(
-                I,
-                &mesh_bind_group.into_inner().normal,
-                &[mesh_index.index()],
-            );
-        }
+        let bind_group = MeshBindGroupId::new;
+
+        let mesh_index = mesh_index.index();
+        let (bind_group_id, indices) = match (skinned_mesh_joints, morph_targets) {
+            (None, None) => (bind_group(false, None), vec![mesh_index]),
+            (Some(joint), None) => (bind_group(true, None), vec![mesh_index, joint.index]),
+            (None, Some(morph)) => (
+                bind_group(false, Some(&morph.image)),
+                vec![mesh_index, morph.index],
+            ),
+            (Some(joint), Some(morph)) => (
+                bind_group(true, Some(&morph.image)),
+                vec![mesh_index, joint.index, morph.index],
+            ),
+        };
+        // TODO: unwrap
+        let Some(bind_group) = prepared_meshes.into_inner().0.get(&bind_group_id) else {
```

Comment:
Shadow variable here for `bind_group` makes the code a bit more complicated to read, given they have very different purposes where one is a function, the other is a bind group.